In [144]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
%matplotlib inline

In [145]:
def categorize(l):
    uniques = sorted(list(set(l)))
    return [uniques.index(x) + 1 for x in l]

In [146]:
# Replace the path with the correct path for your data.
y_2015 = pd.read_csv('/Users/mille/Desktop/Random Forest/LoanStats3d.csv',
    skipinitialspace=True,
    header=1
)

C:\Users\mille\Conda3\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (0,19,55) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [147]:
y_2015.head()

id   member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  68009401  72868139.0    16000.0      16000.0          16000.0   60 months   
1  68354783  73244544.0     9600.0       9600.0           9600.0   36 months   
2  68466916  73356753.0    25000.0      25000.0          25000.0   36 months   
3  68466961  73356799.0    28000.0      28000.0          28000.0   36 months   
4  68495092  73384866.0     8650.0       8650.0           8650.0   36 months   

  int_rate  installment grade sub_grade            ...              \
0   14.85%       379.39     C        C5            ...               
1    7.49%       298.58     A        A4            ...               
2    7.49%       777.55     A        A4            ...               
3    6.49%       858.05     A        A2            ...               
4   19.89%       320.99     E        E3            ...               

  num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75  \
0                0.0                2.0           78.9               0.0   
1                0.0                2.0          100.0              66.7   
2                0.0                0.0          100.0              20.0   
3                0.0                0.0           91.7              22.2   
4                0.0               12.0          100.0              50.0   

  pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort  \
0                  0.0       2.0        298100.0           31329.0   
1                  0.0       0.0         88635.0           55387.0   
2                  0.0       0.0        373572.0           68056.0   
3                  0.0       0.0        304003.0           74920.0   
4                  1.0       0.0         38998.0           18926.0   

  total_bc_limit total_il_high_credit_limit  
0       281300.0                    13400.0  
1        12500.0                    75635.0  
2        38400.0                    82117.0  
3        41500.0                    42503.0  
4         2750.0                    18248.0  

[5 rows x 111 columns]

In [148]:
y_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421097 entries, 0 to 421096
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(85), object(26)
memory usage: 356.6+ MB


In [149]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()
X = y_2015.drop('loan_status', 1)
Y = y_2015['loan_status']
X = pd.get_dummies(X)

cross_val_score(rfc, X, Y, cv=5)

MemoryError: 

In [150]:
categorical = y_2015.select_dtypes(include=['object'])
for i in categorical:
    column = categorical[i]
    print(i)
    print(column.nunique())

id
421097
term
2
int_rate
110
grade
7
sub_grade
35
emp_title
120812
emp_length
11
home_ownership
4
verification_status
3
issue_d
12
loan_status
7
pymnt_plan
1
url
421095
desc
34
purpose
14
title
27
zip_code
914
addr_state
49
earliest_cr_line
668
revol_util
1211
initial_list_status
2
last_pymnt_d
25
next_pymnt_d
4
last_credit_pull_d
26
application_type
2
verification_status_joint
3


In [151]:
# Convert ID and Interest Rate to numeric.
y_2015['id'] = pd.to_numeric(y_2015['id'], errors='coerce')
y_2015['int_rate'] = pd.to_numeric(y_2015['int_rate'].str.strip('%'), errors='coerce')

# Drop other columns with many unique variables
y_2015.drop(['url', 'emp_title', 'zip_code', 'earliest_cr_line', 'revol_util',
            'sub_grade', 'addr_state', 'desc', 'dti', 'emp_length', 'funded_amnt_inv'], 1, inplace=True)

In [152]:
y_2015.tail()

id   member_id  loan_amnt  funded_amnt        term  int_rate  \
421092  36271333.0  38982739.0    13000.0      13000.0   60 months     15.99   
421093  36490806.0  39222577.0    12000.0      12000.0   60 months     19.99   
421094  36271262.0  38982659.0    20000.0      20000.0   36 months     11.99   
421095         NaN         NaN        NaN          NaN         NaN       NaN   
421096         NaN         NaN        NaN          NaN         NaN       NaN   

        installment grade home_ownership  annual_inc  \
421092       316.07     D           RENT     35000.0   
421093       317.86     E           RENT     64400.0   
421094       664.20     B           RENT    100000.0   
421095          NaN   NaN            NaN         NaN   
421096          NaN   NaN            NaN         NaN   

                   ...             num_tl_90g_dpd_24m num_tl_op_past_12m  \
421092             ...                            0.0                3.0   
421093             ...                            1.0                2.0   
421094             ...                            0.0                1.0   
421095             ...                            NaN                NaN   
421096             ...                            NaN                NaN   

       pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens  \
421092          100.0             50.0                  1.0       0.0   
421093           95.0             66.7                  0.0       0.0   
421094          100.0             50.0                  0.0       1.0   
421095            NaN              NaN                  NaN       NaN   
421096            NaN              NaN                  NaN       NaN   

        tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
421092          51239.0            34178.0         10600.0   
421093          96919.0            58418.0          9700.0   
421094          43740.0            33307.0         41700.0   
421095              NaN                NaN             NaN   
421096              NaN                NaN             NaN   

        total_il_high_credit_limit  
421092                     33239.0  
421093                     69919.0  
421094                         0.0  
421095                         NaN  
421096                         NaN  

[5 rows x 100 columns]

In [153]:
# Remove two summary rows at the end that don't actually contain data.
y_2015 = y_2015[:-2]

In [154]:
pd.get_dummies(y_2015)

id   member_id  loan_amnt  funded_amnt  int_rate  installment  \
0       68009401.0  72868139.0    16000.0      16000.0     14.85       379.39   
1       68354783.0  73244544.0     9600.0       9600.0      7.49       298.58   
2       68466916.0  73356753.0    25000.0      25000.0      7.49       777.55   
3       68466961.0  73356799.0    28000.0      28000.0      6.49       858.05   
4       68495092.0  73384866.0     8650.0       8650.0     19.89       320.99   
5       68506798.0  73396623.0    23000.0      23000.0      8.49       471.77   
6       68566886.0  73456723.0    29900.0      29900.0     12.88       678.49   
7       68577849.0  73467703.0    18000.0      18000.0     11.99       400.31   
8       66310712.0  71035433.0    35000.0      35000.0     14.85       829.90   
9       68476807.0  73366655.0    10400.0      10400.0     22.45       289.91   
10      68341763.0  72928789.0    20000.0      20000.0     10.78       432.66   
11      68356421.0  73246236.0    22400.0      22400.0     12.88       508.30   
12      68426545.0  73316365.0    16000.0      16000.0     12.88       363.07   
13      68407277.0  73297111.0     3600.0       3600.0     13.99       123.03   
14      68355089.0  73244858.0    24700.0      24700.0     11.99       820.28   
15      68466926.0  73356763.0    10000.0      10000.0      6.49       306.45   
16      68338832.0  73542831.0     1400.0       1400.0     12.88        47.10   
17      68426831.0  73316657.0    11950.0      11950.0     13.44       405.18   
18      66624733.0  71350547.0    18000.0      18000.0     19.48       471.70   
19      68616873.0  73519699.0     8000.0       8000.0     11.48       263.74   
20      67275481.0  72077199.0    20000.0      20000.0      8.49       631.26   
21      68476668.0  73366513.0    20000.0      20000.0      9.17       637.58   
22      68476702.0  73366548.0    28000.0      28000.0     13.44       949.38   
23      68436666.0  73326493.0     5000.0       5000.0     13.44       169.54   
24      68596180.0  73485987.0    20000.0      20000.0      5.32       602.30   
25      67849662.0  72708407.0     4225.0       4225.0     14.85       146.16   
26      68526883.0  73416719.0    15000.0      15000.0      5.32       451.73   
27      68607141.0  73509983.0    17600.0      17600.0      5.32       530.03   
28      68476715.0  73366561.0     6000.0       6000.0      7.49       186.61   
29      67715283.0  72569997.0    16000.0      16000.0     12.88       538.18   
...            ...         ...        ...          ...       ...          ...   
421065  36631385.0  39374138.0    13000.0      13000.0      8.19       408.52   
421066  36231083.0  38942474.0    15000.0      15000.0     11.99       498.15   
421067  36691397.0  39434133.0    20000.0      20000.0      6.99       617.46   
421068  36743377.0  39486112.0     4200.0       4200.0     15.99       147.64   
421069  36331554.0  39042967.0    26950.0      26950.0      8.19       846.88   
421070  36460697.0  39182359.0    23000.0      23000.0     15.59       554.32   
421071  35773415.0  38468842.0    18700.0      18700.0     17.86       473.44   
421072  36250801.0  38962184.0     6000.0       6000.0     17.86       216.50   
421073  36371802.0  39103231.0    15000.0      15000.0     11.44       329.44   
421074  36440728.0  39152101.0    25000.0      25000.0     15.99       607.82   
421075  36361815.0  39093260.0    26500.0      26500.0      8.67       545.87   
421076  35591131.0  38288343.0    25000.0      25000.0     15.59       602.52   
421077  36201871.0  38913321.0    21000.0      21000.0      6.03       639.15   
421078  36381756.0  39013193.0     8000.0       8000.0      8.19       251.40   
421079  36231724.0  38943173.0    12000.0      12000.0      9.49       384.34   
421080  36301729.0  39023153.0    10000.0      10000.0      6.49       306.45   
421081  36250208.0  38961532.0    13150.0      13150.0     14.99       312.77   
421082  36241316.0  38952731.0     6225.0      

In [155]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()
X = y_2015.drop('loan_status', 1)
Y = y_2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)

cross_val_score(rfc, X, Y, cv=10)

C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: T

array([0.97998148, 0.9801477 , 0.98154876, 0.98176248, 0.96630254,
       0.97919734, 0.95191052, 0.98055048, 0.97964709, 0.98014535])

In [156]:
y_2015.head()

id   member_id  loan_amnt  funded_amnt        term  int_rate  \
0  68009401.0  72868139.0    16000.0      16000.0   60 months     14.85   
1  68354783.0  73244544.0     9600.0       9600.0   36 months      7.49   
2  68466916.0  73356753.0    25000.0      25000.0   36 months      7.49   
3  68466961.0  73356799.0    28000.0      28000.0   36 months      6.49   
4  68495092.0  73384866.0     8650.0       8650.0   36 months     19.89   

   installment grade home_ownership  annual_inc             ...              \
0       379.39     C       MORTGAGE     48000.0             ...               
1       298.58     A       MORTGAGE     60000.0             ...               
2       777.55     A       MORTGAGE    109000.0             ...               
3       858.05     A       MORTGAGE     92000.0             ...               
4       320.99     E           RENT     55000.0             ...               

  num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75  \
0                0.0                2.0           78.9              0.0   
1                0.0                2.0          100.0             66.7   
2                0.0                0.0          100.0             20.0   
3                0.0                0.0           91.7             22.2   
4                0.0               12.0          100.0             50.0   

  pub_rec_bankruptcies tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
0                  0.0       2.0         298100.0            31329.0   
1                  0.0       0.0          88635.0            55387.0   
2                  0.0       0.0         373572.0            68056.0   
3                  0.0       0.0         304003.0            74920.0   
4                  1.0       0.0          38998.0            18926.0   

   total_bc_limit  total_il_high_credit_limit  
0        281300.0                     13400.0  
1         12500.0                     75635.0  
2         38400.0                     82117.0  
3         41500.0                     42503.0  
4          2750.0                     18248.0  

[5 rows x 100 columns]

In [157]:
y_2015 = y_2015.ix[:, 'annual_inc':'total_pymnt']


C:\Users\mille\Conda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


In [158]:
y_2015 = y_2015[:-2]
pd.get_dummies(y_2015)

annual_inc  delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
0          48000.0          0.0             0.0                    33.0   
1          60000.0          0.0             0.0                     NaN   
2         109000.0          0.0             1.0                     NaN   
3          92000.0          0.0             0.0                    42.0   
4          55000.0          0.0             4.0                     NaN   
5          64000.0          0.0             0.0                    29.0   
6          65000.0          0.0             0.0                     NaN   
7         112000.0          0.0             0.0                     NaN   
8         110000.0          0.0             0.0                     NaN   
9         104433.0          1.0             3.0                    12.0   
10         63000.0          0.0             0.0                     NaN   
11         95000.0          0.0             0.0                    54.0   
12         70000.0          0.0             0.0                     NaN   
13         55000.0          0.0             1.0                    30.0   
14         65000.0          1.0             4.0                     6.0   
15         85000.0          0.0             1.0                     NaN   
16         64000.0          0.0             0.0                     NaN   
17         34000.0          0.0             0.0                     NaN   
18        150000.0          0.0             1.0                     NaN   
19         42000.0          0.0             0.0                    75.0   
20         85000.0          1.0             0.0                     3.0   
21        180000.0          0.0             0.0                    49.0   
22        125000.0          0.0             0.0                     NaN   
23         79000.0          0.0             0.0                     NaN   
24         65000.0          0.0             1.0                     NaN   
25         35000.0          2.0             0.0                    18.0   
26         44000.0          0.0             1.0                    47.0   
27         88000.0          0.0             0.0                     NaN   
28        100000.0          1.0             1.0                    15.0   
29         65000.0          0.0             0.0                    33.0   
...            ...          ...             ...                     ...   
421063     63078.0          0.0             0.0                     NaN   
421064     93600.0          1.0             0.0                    17.0   
421065     85000.0          1.0             0.0                    15.0   
421066     70000.0          0.0             0.0                     NaN   
421067     72000.0          0.0             1.0                    69.0   
421068     48000.0          0.0             0.0                    38.0   
421069     54000.0          0.0             0.0                     NaN   
421070     50000.0          0.0             0.0                     NaN   
421071     52000.0          0.0             0.0                    30.0   
421072     90000.0          0.0             0.0                     NaN   
421073     57000.0          0.0             0.0                     NaN   
421074    103000.0          0.0             0.0                     NaN   
421075     92000.0          1.0             0.0                    17.0   
421076     55000.0          0.0             0.0                     NaN   
421077     65000.0          0.0             0.0                     NaN   
421078     50000.0          0.0             1.0                    24.0   
421079     40000.0          1.0             0.0                     8.0   
421080     75000.0          0.0             0.0                    29.0   
421081     30000.0          0.0             0.0                     NaN   
421082     27000.0          0.0             1.0                     NaN   
421083     54000.0          1.0             0.0                    16.0   
421084     57000.0          2.0  

In [159]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()
X = y_2015.drop('loan_status', 1)
Y = y_2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)

cross_val_score(rfc, X, Y, cv=10)

C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\mille\Conda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: T

array([0.92455653, 0.89974116, 0.89978865, 0.91071005, 0.91393968,
       0.90987889, 0.85380797, 0.91954023, 0.91493101, 0.9142165 ])

After a series of dropping and filtering data frame produce a Random Forest of 0.914. Which is very close to 0.90 which was asking in the Thinkful assignment. 

The entire data frame contained 224 columns. To reduce it down to a .914 average value, reduce the columns down to 78. 

Random forest has a tendency to overfit the model. Variables with high variance greatly affect the overfitting. By reducing variables with high variance will provide more accuracy with the model. Leaving in high variance variable will reduce the accuracy in the model. 